include/mbedtls/ccm.h

```diff
@@ -284,10 +289,203 @@ int mbedtls_ccm_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
  */
 int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
                       const unsigned char *iv, size_t iv_len,
-                      const unsigned char *add, size_t add_len,
+                      const unsigned char *ad, size_t ad_len,
                       const unsigned char *input, unsigned char *output,
                       const unsigned char *tag, size_t tag_len );
 
+/**
+ * \brief           This function starts a CCM encryption or decryption
+ *                  operation.
+ *
+ *                  This function and mbedtls_ccm_set_lengths() must be called
+ *                  before calling mbedtls_ccm_update_ad() or
+ *                  mbedtls_ccm_update(). This function can be called before
+ *                  or after mbedtls_ccm_set_lengths().
+ *
+ * \note            This function is not implemented in Mbed TLS yet.
+ *
+ * \param ctx       The CCM context. This must be initialized.
+ * \param mode      The operation to perform: #MBEDTLS_CCM_ENCRYPT or
+ *                  #MBEDTLS_CCM_DECRYPT or #MBEDTLS_CCM_STAR_ENCRYPT or
+ *                  #MBEDTLS_CCM_STAR_DECRYPT.
+ * \param iv        The initialization vector. This must be a readable buffer
+ *                  of at least \p iv_len Bytes.
+ * \param iv_len    The length of the nonce in Bytes: 7, 8, 9, 10, 11, 12,
+ *                  or 13. The length L of the message length field is
+ *                  15 - \p iv_len.
+ *
+ * \return          \c 0 on success.
+ * \return          \#MBEDTLS_ERR_CCM_BAD_INPUT on failure:
+ *                  \p mode is invalid,
```

Comment:
A context in an invalid state is also a possible reason for `MBEDTLS_ERR_CCM_BAD_INPUT` and we should document it. We just need to be careful not to mandate detecting all possible ways the context can be invalid. We can just say “`ctx` is in an invalid state” without details.

Reply:
I thought about that but didn't go that way because:
1) looking at the code of mbedtls_gcm_starts() it does not seem that we want to enforce this in the sense of checking the context state at the beginning of the function. If mbedtls_gcm_starts() is called with a context in an invalid state it will fail (hopefully) with an error code but not sure which one.
2) if we want to enforce that mbedtls_gcm_starts() is called with a context in a valid state then I'd say that MBEDTLS_ERR_CCM_BAD_INPUT is not really a good error code, _BAD_STATE is probably better.

Having said that, looking at the GCM implementation it seems possible for MBEDTLS_ERR_CCM_BAD_INPUT to be returned if the context is in an invalid state thus I can add it if you still prefer. I guess that in that case we would need to add "ctx is in an invalid state" in the documentation of the other new APIs as well, right?

Reply:
A lot of old Mbed TLS documentation has an incomplete description of error conditions. I prefer to be complete in new documentation. We shouldn't promise that all cases of invalid contexts will be detected, but we should document that if the function returns `BAD_INPUT`, it can mean that it's been called with a context in an invalid state.

There's no `BAD_STATE` error in any mbedtls module except chachapoly. Others return `BAD_INPUT_DATA` for a context in a bad state, even high-level modules like cipher.

Reply:
done